File: frontend/src/data/checksum.ts
import {useEffect, useRef, useState} from 'preact/hooks';
import {PlaysetChecksum} from '../../wailsjs/go/main/App';
import type {main} from '../../wailsjs/go/models';

// The multiplayer checksum: the four characters a Paradox game shows on its main menu, which
// every player in a multiplayer game has to share. The backend (internal/checksum, checksum.go)
// works it out from the saved playset without starting the game; this is the interface half -
// when to ask, and how to say what came back.

export interface ChecksumState {
    // 'none': nothing to show (asked for nothing yet, or a game with no known scheme).
    kind: 'none' | 'calculating' | 'ready' | 'unavailable';
    // The saved playset it was asked for.
    forName: string;
    // The code, files and mods that went into it - while calculating, those of the previous
    // answer, so the number does not flicker away and back for a calculation that takes a moment.
    value: string;
    files: number;
    mods: number;
    // Why there is no code, in words for the user.
    reason: string;
    warnings: string[];
}

export const NO_CHECKSUM: ChecksumState = {kind: 'none', forName: '', value: '', files: 0, mods: 0, reason: '', warnings: []};

function fromResult(name: string, r: main.PlaysetChecksumResult): ChecksumState {
    if (r.Status === 'ready') {
        return {kind: 'ready', forName: name, value: r.Checksum, files: r.Files, mods: r.Mods, reason: '', warnings: r.Warnings ?? []};
    }
    if (r.Status === 'unavailable') {
        return {...NO_CHECKSUM, kind: 'unavailable', forName: name, reason: r.Reason};
    }
    return NO_CHECKSUM;
}

export interface PlaysetChecksumControl {
    state: ChecksumState;
    // Works out the checksum of the saved playset called name (after it was saved or loaded).
    calculate(name: string): void;
    // Works it out again for the last playset asked about, e.g. after mods changed on disk.
    refresh(): void;
}

// usePlaysetChecksum keeps the checksum of gameId's playset. A newer request replaces an older
// one still running (only the latest answer is used), and switching game clears it.
export function usePlaysetChecksum(gameId: string): PlaysetChecksumControl {
    const [state, setState] = useState<ChecksumState>(NO_CHECKSUM);
    const latest = useRef(0);
    const lastName = useRef('');
    const gameRef = useRef(gameId);
    gameRef.current = gameId;

    useEffect(() => {
        latest.current++;
        lastName.current = '';
        setState(NO_CHECKSUM);
    }, [gameId]);

    const control = useRef<PlaysetChecksumControl | null>(null);
    if (control.current === null) {
        control.current = {
            state,
            calculate(name: string) {
                const trimmed = name.trim();
                if (!trimmed || !gameRef.current) return;
                const mine = ++latest.current;
                lastName.current = trimmed;
                setState((prev) => (prev.kind === 'ready' && prev.forName === trimmed
                    ? prev
                    : {...NO_CHECKSUM, kind: 'calculating', forName: trimmed}));
                PlaysetChecksum(gameRef.current, trimmed).then((r) => {
                    if (mine === latest.current) setState(fromResult(trimmed, r));
                }).catch((err) => {
                    if (mine === latest.current) setState({...NO_CHECKSUM, kind: 'unavailable', forName: trimmed, reason: String(err)});
                });
            },
            refresh() {
                if (lastName.current) control.current!.calculate(lastName.current);
            },
        };
    }
    control.current.state = state;
    return control.current;
}

// Whether the checksum is worth showing for the playset on screen: only for the playset it was
// worked out for (a renamed, deleted or new one has none yet).
export function checksumShown(state: ChecksumState, playsetName: string): boolean {
    return state.kind !== 'none' && state.forName === playsetName.trim();
}

// What the small checksum readouts (above Play, in the launch dialog) show.
export interface ChecksumBadge {
    // The code, or a word standing in for it.
    text: string;
    // 'ok' a code to share, 'busy' still working, 'warn' none to give (or one about to change).
    tone: 'ok' | 'busy' | 'warn';
}

// checksumBadge is null when there is nothing to show for the playset on screen.
export function checksumBadge(state: ChecksumState, playsetName: string, unsaved: boolean): ChecksumBadge | null {
    if (!checksumShown(state, playsetName)) return null;
    if (state.kind === 'calculating') return {text: 'calculating...', tone: 'busy'};
    if (state.kind === 'unavailable') return {text: 'unavailable', tone: 'warn'};
    return {text: state.value, tone: unsaved ? 'warn' : 'ok'};
}

File: internal/app/checksum.go
package app

import (
	"context"
	"errors"
	"fmt"
	"path"
	"path/filepath"
	"strings"
	"time"

	"github.com/Official-Husko/parallax-mod-manager/internal/applog"
	"github.com/Official-Husko/parallax-mod-manager/internal/checksum"
	"github.com/Official-Husko/parallax-mod-manager/internal/game"
	"github.com/Official-Husko/parallax-mod-manager/internal/mod"
	"github.com/Official-Husko/parallax-mod-manager/internal/scan"
)

// Checksum statuses, as PlaysetChecksum reports them.
const (
	// checksumReady: Checksum holds the calculated code.
	checksumReady = "ready"
	// checksumUnavailable: it could not be calculated; Reason says why.
	checksumUnavailable = "unavailable"
	// checksumUnsupported: this game has no known checksum scheme, so there is nothing to show.
	checksumUnsupported = "unsupported"
)

// PlaysetChecksumResult is the multiplayer checksum of a saved playset.
type PlaysetChecksumResult struct {
	Status string
	// Checksum is the four characters the game shows on its main menu.
	Checksum string
	// Files is how many files went into it, Mods how many enabled mods.
	Files int
	Mods  int
	// Reason says, in words for the user, why there is no checksum.
	Reason string
	// Warnings are things worth knowing that did not stop the calculation.
	Warnings []string
}

// PlaysetChecksum calculates the checksum the game will show for the named saved playset,
// without starting the game: the code every player in a multiplayer game has to share.
//
// It describes the playset as saved (what launching loads), not unsaved edits, and is meant to be
// asked for again whenever the playset is saved or loaded. A newer request for the same game
// stops an older one still running; the stopped one returns an error the caller can ignore.
func (a *App) PlaysetChecksum(gameID, playsetName string) (PlaysetChecksumResult, error) {
	cfg, ok := a.registry.Get(gameID)
	if !ok {
		return PlaysetChecksumResult{}, fmt.Errorf("app: unknown game %q", gameID)
	}
	if !checksum.Algorithm(cfg.ChecksumAlgorithm).Known() {
		return PlaysetChecksumResult{Status: checksumUnsupported}, nil
	}

	ctx, done := a.startChecksum(gameID)
	defer done()

	log := applog.For("Checksum")
	started := time.Now()
	res, err := a.playsetChecksum(ctx, cfg, playsetName)
	if err != nil {
		if ctx.Err() != nil {
			return PlaysetChecksumResult{}, ctx.Err()
		}
		return PlaysetChecksumResult{}, err
	}
	took := time.Since(started).Round(time.Millisecond)
	switch res.Status {
	case checksumReady:
		log.Infof("'%s' with playset '%s': %s (%d files, %d mods, %s)", cfg.DisplayName, playsetName, res.Checksum, res.Files, res.Mods, took)
		for _, w := range res.Warnings {
			log.Warnf("%s", w)
		}
	default:
		log.Warnf("'%s' with playset '%s': no checksum - %s", cfg.DisplayName, playsetName, res.Reason)
	}
	return res, nil
}

// startChecksum registers a new calculation for gameID, cancelling the one still running for
// it. done releases the registration.
func (a *App) startChecksum(gameID string) (context.Context, func()) {
	ctx, cancel := context.WithCancel(a.baseContext())
	a.checksumMu.Lock()
	if a.checksumCancel == nil {
		a.checksumCancel = map[string]context.CancelFunc{}
	}
	if prev := a.checksumCancel[gameID]; prev != nil {
		prev()
	}
	a.checksumCancel[gameID] = cancel
	a.checksumMu.Unlock()
	return ctx, cancel
}

// playsetChecksum does the work of PlaysetChecksum for a game with a known scheme.
func (a *App) playsetChecksum(ctx context.Context, cfg game.GameConfig, playsetName string) (PlaysetChecksumResult, error) {
	unavailable := func(format string, args ...any) (PlaysetChecksumResult, error) {
		return PlaysetChecksumResult{Status: checksumUnavailable, Reason: fmt.Sprintf(format, args...)}, nil
	}
	if strings.TrimSpace(playsetName) == "" {
		return unavailable("No playset is loaded.")
	}
	p, err := a.playsets.Load(ctx, cfg.ID, playsetName)
	if err != nil {
		return PlaysetChecksumResult{}, err
	}
	installDir, ok := a.resolveInstallDir(cfg)
	if !ok {
		return unavailable("%s's install folder was not found.", cfg.DisplayName)
	}
	userDir, err := cfg.UserDataDir()
	if err != nil {
		return PlaysetChecksumResult{}, err
	}
	modDir := filepath.Join(userDir, "mod")

	scanned, err := scan.Scan(ctx, scan.Options{Game: cfg, SteamRoots: a.steamRoots, ExtraFolders: a.extraModFolders(cfg.ID)})
	if err != nil {
		return PlaysetChecksumResult{}, err
	}
	return computePlaysetChecksum(ctx, cfg, installDir, modDir, p.ModIDs, scanned.Mods, &a.checksumCache)
}

// computePlaysetChecksum calculates the checksum of the game with the mods named by ids (in
// load order) laid over it. Anything that makes the result untrustworthy - a mod that is not
// installed, one whose folder is missing - is reported as unavailable rather than left out.
func computePlaysetChecksum(ctx context.Context, cfg game.GameConfig, installDir, modDir string, ids []string, scanned []mod.Mod, cache *checksum.ResultCache) (PlaysetChecksumResult, error) {
	byID := make(map[string]mod.Mod, len(scanned))
	for _, m := range scanned {
		byID[m.ID] = m
	}

	var mods []checksum.Mod
	var notInstalled []string
	seen := map[string]bool{}
	for _, id := range ids {
		if seen[id] {
			continue
		}
		seen[id] = true
		m, ok := byID[id]
		if !ok {
			notInstalled = append(notInstalled, id)
			continue
		}
		mods = append(mods, checksum.Mod{
			ID:           m.ID,
			Name:         m.Descriptor.Name,
			RegistryID:   path.Join("mod", m.ID+".mod"),
			Content:      contentForChecksum(m),
			ReplacePaths: m.Descriptor.ReplacePath,
			Dependencies: m.Descriptor.Dependencies,
		})
	}
	if len(notInstalled) > 0 {
		return PlaysetChecksumResult{
			Status: checksumUnavailable,
			Reason: fmt.Sprintf("%s in this playset %s not installed: %s.", countOf(len(notInstalled), "mod"), isAre(len(notInstalled)), strings.Join(notInstalled, ", ")),
		}, nil
	}

	res, err := checksum.Compute(ctx, checksum.Input{
		Algorithm:        checksum.Algorithm(cfg.ChecksumAlgorithm),
		GameDir:          installDir,
		LauncherSettings: filepath.Join(installDir, filepath.FromSlash(cfg.LauncherSettingsPath)),
		ModDir:           modDir,
		Mods:             mods,
		Cache:            cache,
	})
	var missing *checksum.MissingContentError
	switch {
	case errors.As(err, &missing):
		return PlaysetChecksumResult{
			Status: checksumUnavailable,
			Reason: fmt.Sprintf("%s in this playset %s no files on disk: %s.", countOf(len(missing.Mods), "mod"), hasHave(len(missing.Mods)), strings.Join(missing.Mods, ", ")),
		}, nil
	case err != nil:
		return PlaysetChecksumResult{}, err
	}
	return PlaysetChecksumResult{
		Status:   checksumReady,
		Checksum: res.Checksum,
		Files:    res.Files,
		Mods:     len(res.Order),
		Warnings: res.Warnings,
	}, nil
}

// contentForChecksum is where the game finds m's files. A mod whose folder was not found has
// none, which the calculation reports rather than skips.
func contentForChecksum(m mod.Mod) string {
	if m.ContentMissing {
		return ""
	}
	return m.ContentPath
}

func countOf(n int, noun string) string {
	if n == 1 {
		return "1 " + noun
	}
	return fmt.Sprintf("%d %ss", n, noun)
}

func isAre(n int) string {
	if n == 1 {
		return "is"
	}
	return "are"
}

func hasHave(n int) string {
	if n == 1 {
		return "has"
	}
	return "have"
}
